pkg/jobrunaggregator/jobrunaggregatoranalyzer/analyzer_disruption.go

```diff
@@ -0,0 +1,174 @@
+	"github.com/openshift/ci-tools/pkg/junit"
+)
+
+func (o *JobRunAggregatorAnalyzerOptions) CalculateDisruptionTestSuite(ctx context.Context, jobName string, finishedJobsToAggregate []jobrunaggregatorapi.JobRunInfo) (*junit.TestSuite, error) {
+	disruptionJunitSuite := &junit.TestSuite{
+		Name:      "BackendDisruption",
+		TestCases: []*junit.TestCase{},
+	}
+	collectedDataTestCase := &junit.TestCase{
+		Name: "should collect disruption data",
+	}
+	disruptionJunitSuite.TestCases = append(disruptionJunitSuite.TestCases, collectedDataTestCase)
+
+	jobRunIDToBackendNameToAvailabilityResult, err := getDisruptionByJobRunID(ctx, finishedJobsToAggregate)
+	if jobRunIDToBackendNameToAvailabilityResult != nil {
+		rawDataBytes, err := yaml.Marshal(jobRunIDToBackendNameToAvailabilityResult)
+		if err != nil {
+			collectedDataTestCase.SystemOut = string(rawDataBytes)
+		}
+	}
+	switch {
+	case len(jobRunIDToBackendNameToAvailabilityResult) < 3 && err != nil:
+		return nil, err
+	case len(jobRunIDToBackendNameToAvailabilityResult) < 3 && err == nil:
+		collectedDataTestCase.FailureOutput = &junit.FailureOutput{
+			Message: "not enough data to aggregate",
+			Output:  collectedDataTestCase.SystemOut,
+		}
+		disruptionJunitSuite.NumFailed++
+		return disruptionJunitSuite, nil
+
+	default:
+		// ignore the errors if we have at least three results
+		fmt.Fprintf(os.Stderr, "Could not fetch backend disruption data for all runs %v\n", err)
+	}
+
+	allBackends := getAllDisruptionBackendNames(jobRunIDToBackendNameToAvailabilityResult)
+	for _, backendName := range allBackends.List() {
+		jobRunIDToAvailabilityResultForBackend := getDisruptionForBackend(jobRunIDToBackendNameToAvailabilityResult, backendName)
+		historicalStats, failed, message, err := o.passFailCalculator.CheckDisruption(ctx, jobRunIDToAvailabilityResultForBackend, backendName)
+		if err != nil {
+			return nil, err
+		}
+
+		junitTestCase := &junit.TestCase{
+			Name: fmt.Sprintf("%s should remain available", backendName),
+		}
+		disruptionJunitSuite.TestCases = append(disruptionJunitSuite.TestCases, junitTestCase)
+
+		currDetails := TestCaseDetails{
+			Name:    junitTestCase.Name,
+			Summary: message,
+		}
+		for jobRunID := range jobRunIDToAvailabilityResultForBackend {
+			currAvailabilityStat := jobRunIDToAvailabilityResultForBackend[jobRunID]
+			humanURL := jobrunaggregatorapi.GetHumanURL(jobName, jobRunID)
+			gcsArtifactURL := jobrunaggregatorapi.GetGCSArtifactURL(jobName, jobRunID)
+			overMean := float64(currAvailabilityStat.SecondsUnavailable) > historicalStats.mean
+			overP95 := float64(currAvailabilityStat.SecondsUnavailable) > historicalStats.p95
+			switch {
+			case overP95:
+				currDetails.Failures = append(currDetails.Failures, TestCaseFailure{
+					JobRunID:       jobRunID,
+					HumanURL:       humanURL,
+					GCSArtifactURL: gcsArtifactURL,
+				})
+
+			case overMean: // this will mark it as a flake in higher layers
```

Comment:
Only time we'll fail is if the mean for the job run is over the historical p95?

I was marking any backend that was more than 110% of the mean, excluding 1 outlier, as a failure.